Imgcustomizer: Discover partitions using grub.cfg and fstab files

cwize1 opened the pull request:
###### Merge Checklist 
**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary

In the imgcustomizer tool, detect the partitions available on the disk and how they should be mounted by reading the `grub.cfg` file on the boot partition and the `/etc/fstab` file on the OS partition. This will allow the tool to work on a variety of images with custom partition layouts.

###### Change Log

- In imgcustomizer, detect partitions using `grub.cfg` and `fstab` files.
- Update some `installutils` functions so that they are usable by imgcustomizer.
- Ensure that the assets directory mount point is only specified once within the code.

###### Does this affect the toolchain?

NO

###### Test Methodology

- Updated the UTs for imgcustomizer.

romoh commented on toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go:
```diff
@@ -158,14 +168,134 @@ func customizeImageHelper(buildDir string, baseConfigPath string, config *imagec
+	// Look for the boot partition (i.e. EFI system partition).
+	var efiSystemPartition *diskutils.PartitionInfo
+	for _, diskPartition := range diskPartitions {
+		if diskPartition.PartitionTypeUuid == "c12a7328-f81f-11d2-ba4b-00a0c93ec93b" {
```
move to const

dmcilvaney commented on toolkit/tools/internal/safemount.go/safemount.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+// Creates a new system mount.
+func NewMount(source, target, fstype string, flags uintptr, data string, makeAndDeleteDir bool) (*Mount, error) {
```
Thoughts on adding a bit more smarts here to coordinate nested mounts/unmounts (ie `/boot`, `/boot/efi`)? Something I've had bubbling in the back of my brain is to create a standardized way to manage mount points in the imager tool that knows how to cleanup after itself. Not critical for "simple" mounts, but dm-verity and dm-crypt mounts can get a bit complicated.

cwize1 replied:
The general rule of thumb is to cleanup in the reverse of the order of setup. Would this be sufficient for dm-verity and dm-crypt?

It might make sense to create a safemount list type. Though, I don't need such a thing in this change.

dmcilvaney replied:
Reverse order may be sufficient, I'm mostly interested in adding smarts to each 'type' of mount that knows their specific cleanup requirements. This is mostly a commentary on how disorganized the imager tool is though.

No need to do this now though, we can revisit later.

dmcilvaney reviewed (commented):
Could we also do one of two things:
1) Split foundational stuff into its own PR
2) Promise to do a rebase/cleanup + non-squash merge to preserve the changes as separate commits.

cwize1 commented on toolkit/tools/imagegen/configuration/systemconfig.go:
```diff
@@ -53,18 +53,6 @@ func (s *SystemConfig) GetRootPartitionSetting() (rootPartitionSetting *Partitio
 	return nil
```
I changed the function to take a `[]PartitionSetting` instead of a `*SystemConfig`. I figured it made sense to move the location of the function since it was no longer related to `SystemConfig`.

cwize1 commented on toolkit/tools/imagegen/diskutils/fstab.go:
```diff
@@ -0,0 +1,147 @@
+	"golang.org/x/sys/unix"
+)
+
+type MountFlags uintptr
```
The [unix.Mount](https://pkg.go.dev/golang.org/x/sys/unix#Mount) functions takes `uintptr`.

cwize1 commented on toolkit/tools/imagegen/installutils/installutils.go:
```diff
@@ -962,15 +963,16 @@ func InstallGrubEnv(installRoot string) (err error) {
 // Note: this boot partition could be different than the boot partition specified in the bootloader.
 // This boot partition specifically indicates where to find the kernel, config files, and initrd
-func InstallGrubCfg(installRoot, rootDevice, bootUUID, bootPrefix string, encryptedRoot diskutils.EncryptedRootDevice, kernelCommandLine configuration.KernelCommandLine, readOnlyRoot diskutils.VerityDevice, isBootPartitionSeparate bool) (err error) {
+func InstallGrubCfg(installRoot, rootDevice, bootUUID, bootPrefix, assetsDir string, encryptedRoot diskutils.EncryptedRootDevice, kernelCommandLine configuration.KernelCommandLine, readOnlyRoot diskutils.VerityDevice, isBootPartitionSeparate bool) (err error) {
```
Thanks for the heads up.

cwize1 replied:
I've rebased against that change.

cwize1 commented on toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go:
```diff
@@ -158,14 +168,134 @@ func customizeImageHelper(buildDir string, baseConfigPath string, config *imagec
+	// Mount the boot partition.
+	tmpDir := filepath.Join(buildDir, "tmppartition")
+
+	efiSystemPartitionMount, err := safemount.NewMount(efiSystemPartition.Path, tmpDir, efiSystemPartition.FileSystemType, 0, "", true)
```
Yeah, my intention is for this type to be generically useful.

cwize1 commented on toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go:
```diff
@@ -158,14 +168,134 @@ func customizeImageHelper(buildDir string, baseConfigPath string, config *imagec
+	}
+
+	if !foundRoot {
+		return nil, nil, fmt.Errorf("image has invalid fstab file: no root partition found")
```
Yeah, I fully expect that to be true. My expectation is that this code will expand over time to support more and more different types of images.

cwize1 commented:
> Could we also do one of two things:
> 
> 1. Split foundational stuff into its own PR
> 2. Promise to do a rebase/cleanup + non-squash merge to preserve the changes as separate commits.
>    (maybe add new mount function, move partition query functions / refactor grub assets, etc...)

Is it mainly the assets directory stuff you want in a separate change?

cwize1 commented:
> > Could we also do one of two things:
> > 
> > 1. Split foundational stuff into its own PR
> > 2. Promise to do a rebase/cleanup + non-squash merge to preserve the changes as separate commits.
> >    (maybe add new mount function, move partition query functions / refactor grub assets, etc...)
> 
> Is it mainly the assets directory stuff you want in a separate change?

I've split off the assets directory change: https://github.com/microsoft/CBL-Mariner/pull/6282

cwize1 commented on toolkit/tools/imagegen/diskutils/diskutils.go:
```diff
@@ -674,6 +690,28 @@ func SystemBlockDevices() (systemDevices []SystemBlockDevice, err error) {
+	}
+
+	// Read the disk's partitions.
+	jsonString, _, err := shell.Execute("lsblk", diskDevPath, "--output", "NAME,PATH,PARTTYPE,FSTYPE,UUID,MOUNTPOINT,PARTUUID", "--json", "--list")
```
Fixed.

cwize1 commented on toolkit/tools/imagegen/diskutils/diskutils.go:
```diff
@@ -674,6 +690,28 @@ func SystemBlockDevices() (systemDevices []SystemBlockDevice, err error) {
 	return
 }
 
+func GetDiskPartitions(diskDevPath string) ([]PartitionInfo, error) {
```
Foxed/

cwize1 commented on toolkit/tools/imagegen/diskutils/diskutils.go:
```diff
@@ -674,6 +690,28 @@ func SystemBlockDevices() (systemDevices []SystemBlockDevice, err error) {
+	// Read the disk's partitions.
+	jsonString, _, err := shell.Execute("lsblk", diskDevPath, "--output", "NAME,PATH,PARTTYPE,FSTYPE,UUID,MOUNTPOINT,PARTUUID", "--json", "--list")
+	if err != nil {
+		return nil, fmt.Errorf("failed to list disk (%s) partitions: %w", diskDevPath, err)
```
Fixed.

cwize1 commented on toolkit/tools/imagegen/diskutils/fstab.go:
```diff
@@ -0,0 +1,147 @@
+	// Read the fstab file.
+	// The `findmnt` command provides a convenient JSON output. In addition, it helpfully splits the
+	// common vfs options from the filesystem specific options.
+	jsonString, _, err := shell.Execute("findmnt", "--fstab", "--tab-file", fstabPath,
```
This isn't currently being used by the imager. So, it isn't actually a prerequisite (yet).

Also, if the imager ever does want it, then I suspect it will use it from the setup chroot. So, it would need to be added to the setup chroot, instead of being a host prerequisite.

(I also plan on adding the setup chroot to the image customizer tool. But that'll be done in a later change.)

dmcilvaney replied:
Sure, that seems reasonable.
